confused-Techie/atom-backend
# fix postgre begin() + more tests + code refactoring

### Requirements 

* [x] Have you ran tests against this code?

### Description of the Change

Related to #91. This contains all the commits of #97 plus an additional one which shows an issue.

Which are the steps of the tests:

1. We have a starting package with 1.0.0 version.

2. Add a new 1.0.1 version. OK 

3. Insert a duplicate 1.0.1 version. This has to fail. OK
- - This fails because we compare the semver. If it's lower or equal, we exit.

- In the current state we have:
- - version 1.0.0 published
- - version 1.0.1 latest 

4. Remove 1.0.1 version. OK
- - With checks we perform, the version is marked as removed and the remaining highest published one is converted as latest.

- In the current state we have:
- - version 1.0.0 latest
- - version 1.0.1 removed

5. Check if 1.0.0 is the latest. OK

6. Insert a previous deleted package. Try to reinsert 1.0.1 version. This has to fail. OK
- -  Why we want this to fail. @confused-Techie citation: 
> if I download version 1.0.1 but there's some critical bug or secret within it, and the author pulls it off the store, then uploads again, if they kept the same version, I would personally never see an update and keep running this insecure, secret riddled version
- - This won't fail on semver comparison because the version is higher. That's why we added the (package, semver) unique constraint to prevent this insertion.

7. **The previous step fails as we want but the rollback is not applied!**
- - Before inserting the new version, we convert the current latest version to published status and this operation is saved. If the rollback worked, this was not saved.

- In the current state we have:
- - version 1.0.0 published
- - version 1.0.1 removed 

8. Show the log of `latestVer.content.versions`. We clearly see the only non removed version is published, not latest.

9. Insert a new 1.1.0 version. FAILS
- - Indeed this fails because there are no latest versions as we check inside `insertNewPackageVersion`. So this confirm the rollback is not applied. 

So I wonder why it's not working. I read the **[doc](https://github.com/porsager/postgres#readme)** of the module and it seems that whenever we go into the catch statement, the rollback should be performed. I also applied "`read write`" options, but nothing changes.

This has to be investigated because we need transactions and in the current state begin function is useless.
